### PR TITLE
VZ-3934.  Create effective CR using profiles and input spec

### PR DESCRIPTION
**Pull request by michael-cico**
# Description

Create the effective Verrazzano configuration using profiles and input spec, for platform-operator installs of components.  This utilizes the strategic merge capabilities added earlier in the `controllers/verrazzano/transform` package in the platform operator to merge profiles with user overrides in the declared Verrazzano CR instance (the "actual" CR) and create a merged  CR (the "effective" CR).

This is needed to replicate/replace similar merge/override functionality in the install job scripts, which are being phased out as we move component lifecycle management into the platform operator.  Components will utilize the `spi.ComponentContext.EffectiveCR()` to drive installs/updates/upgrades.

Changes:
- Create profile YAML files in the `platform-operator/manifests/profiles` directory for each of the supported profiles
- Updated the Dockerfile to copy the manifests dir into the image
- Annotate the merge strategy for VZ types that need it; OCI DNS, InstallArgs, Volume overrides, and certificates
- Implement the creation of the `EffectiveCR` property of the `spi.ComponentContext` object using the `transform` package utilities
- Add profile merge unit tests
- Fixed a bug in the strategic merge helpers and updated the those unit tests
- Update Kiali component remove hard-coded check for managed-cluster profile, rely on EffectiveCR for enabled value
- Update Kiali unit tests, add enabled checks for various profile types
- Added a new spi.NewFakeContext test utility to simplify using `ComponentContext` in component unit tests, and update the unit tests accordingly

Fixes VZ-3934.

# Checklist 

As the author of this PR, I have:

- [x] Checked that I included or updated copyright and license notices in all files that I altered
- [x] Added or updated unit tests for any new functions I added
- [ ] Added or updated integration tests if appropriate
- [ ] Added or updated acceptance tests if appropriate

Code reviewer, please confirm this PR:

- [ ] Addressed the requirement and meets the acceptance criteria
- [ ] Does not introduce unrelated or spurious changes
- [ ] Does not introduce any unapproved dependency
- [ ] Makes sense and it easy to understand, and/or difficult areas of code are clearly documented so that they can be understood

### CODE REVIEW

File: platform-operator/apis/verrazzano/v1alpha1/verrazzano_types.go [hunk 1]
```diff
@@ -76,11 +76,13 @@ type VerrazzanoSpec struct {
...
 	// is used, it must reference a VolumeClaimSpecTemplate in the VolumeClaimSpecTemplates section.
 	// +optional
-	DefaultVolumeSource *corev1.VolumeSource `json:"defaultVolumeSource,omitempty"`
+	// +patchStrategy=replace
+	DefaultVolumeSource *corev1.VolumeSource `json:"defaultVolumeSource,omitempty" patchStrategy:"replace"`
 
 	// VolumeClaimSpecTemplates Defines a named set of PVC configurations that can be referenced from components using persistent volumes.
 	// +optional
-	VolumeClaimSpecTemplates []VolumeClaimSpecTemplate `json:"volumeClaimSpecTemplates,omitempty"`
+	// +patchStrategy=replace
```
Comment (michael-cico) [inline platform-operator/apis/verrazzano/v1alpha1/verrazzano_types.go:84]: Might want to do a `merge,retainKeys` on this

Comment (michael-cico) [inline platform-operator/apis/verrazzano/v1alpha1/verrazzano_types.go:84]: change to merge?

File: platform-operator/apis/verrazzano/v1alpha1/verrazzano_types.go [hunk 2]
```diff
@@ -448,7 +464,8 @@ type InstallArgs struct {
 	SetString bool `json:"setString,omitempty"`
 	// List of values for named install argument
 	// +optional
-	ValueList []string `json:"valueList,omitempty"`
+	// +patchStrategy=replace
```
Comment (michael-cico) [inline platform-operator/apis/verrazzano/v1alpha1/verrazzano_types.go:467]: merge

File: platform-operator/controllers/verrazzano/component/spi/component_context.go [hunk 3]
```diff
@@ -6,19 +6,86 @@ package spi
...
+		dryRun:      dryRun,
+		cr:          actualCR,
+		effectiveCR: effectiveCR,
+	}
+}
+
+// getEffectiveCR Creates an "effective" Verrazzano CR based on the user defined resource merged with the profile definitions
+// - Effective CR == base profile + declared profiles + ActualCR (in order)
+// - last definition wins
+func getEffectiveCR(actualCR *vzapi.Verrazzano) (*vzapi.Verrazzano, error) {
```
Comment (michael-cico) [inline platform-operator/controllers/verrazzano/component/spi/component_context.go:69]: Move to `transform` package?

File: platform-operator/apis/verrazzano/v1alpha1/verrazzano_types.go (hunk 1 above)
Comment (pfmackin) [reply platform-operator/apis/verrazzano/v1alpha1/verrazzano_types.go:84]: If we have VolumeClaimSpecTemplates in our profiles and the user defines a new one they should be merged.

File: platform-operator/controllers/verrazzano/component/spi/component_context.go (hunk 3 above)
Comment (pfmackin) [reply platform-operator/controllers/verrazzano/component/spi/component_context.go:69]: I think it might be better in transform,  maybe effective_cr.go

**Review by pfmackin — APPROVED** [review]

File: platform-operator/controllers/verrazzano/component/spi/component_context.go (hunk 3 above)
Comment (michael-cico) [reply platform-operator/controllers/verrazzano/component/spi/component_context.go:69]: Will fix on next pass.

File: platform-operator/apis/verrazzano/v1alpha1/verrazzano_types.go (hunk 1 above)
Comment (michael-cico) [reply platform-operator/apis/verrazzano/v1alpha1/verrazzano_types.go:84]: Will revisit next pass.

**Pull request merged**